timemachine/fe/reweighting.py

```diff
@@ -0,0 +1,309 @@
+    return logsumexp(log_values - np.log(len(log_values)))
+
+
+def estimate_log_z_ratio(log_importance_weights: Array) -> float:
+    """stable log(mean(importance_weights))"""
+    return log_mean(log_importance_weights)
+
+
+def one_sided_exp(delta_us: Array) -> float:
+    """exponential averaging
+
+    References
+    ----------
+    [1] pymbar implementation
+        https://github.com/choderalab/pymbar/blob/15f932a271343e611ed4be2d468c77b1d11cf01f/pymbar/exp.py#L54
+    """
+    # delta_us = -log_importance_weights
+    # delta_f  = -log_z_ratio
+    return -estimate_log_z_ratio(-delta_us)
+
+
+def interpret_as_mixture_potential(u_kn: Array, f_k: Array, N_k: Array) -> Array:
+    r"""Interpret samples from multiple states k as if they originate from a single state
+    defined as a weighted mixture:
+
+    $p_{mix}(x) \propto \sum_k w_k q_k(x) / Z_k$
+
+    where
+    * $q_k(x) = exp(-u_k(x))$ is the Boltzmann weight function for state k
+    * $f_k = - log Z_k$ is the assumed normalization for state k
+    * $w_k \propto N_k$ is the mixture weight of state k
+
+    (up to a single constant)
+
+    Parameters
+    ----------
+    u_kn : [K, N] float array
+        reduced potentials of all N samples evaluated in all K states
+        u_kn[k, n] = u_k(x_n)
+    f_k : [K,] float array
+        reduced free energies of all K states
+        (up to an additive constant)
+    N_k : [K,] int array
+        number of samples from each individual state
+        (sum(N_k) must equal N)
+
+    Returns
+    -------
+    mixture_u_n : [N,] float array
+        mixture_u_n[n] = u_mix(x_n), where
+
+        u_mix(x) = -log(q_mix(x))
+        q_mix(x) = \sum_k w_k (q_k(x) / Z_k)
+
+    Notes
+    -----
+    * This is nicely illustrated in ref [1] as a way to interpret MBAR [2]
+    * Depends on the accuracy of the input free energy estimates (f_k - f_k[0]) for the source states.
+    * This is not the only way to interpret samples from multiple distributions
+        as if they come from a single mixture distribution.
+        In ref [3], alternatives are systematically enumerated.
+        Assuming f_k are exact, the current approach corresponds to N3 in ref [3].
+        (N3 the most computationally expensive of the options in ref [3],
+        since it requires evaluating all k energy functions on every sample,
+        but it is also the lowest variance.)
+
+    References
+    ----------
+    [1] [Shirts, 2017] Reweighting from the mixture distribution as a better way to describe
+        the Multistate Bennett Acceptance Ratio
+        https://arxiv.org/abs/1704.00891
+    [2] [Shirts, Chodera, 2008] Statistically optimal analysis of samples from multiple equilibrium states.
+        J. Chem. Phys. 129:124105, 2008.
+        http://dx.doi.org/10.1063/1.2978177
+    [3] [Elvira+, 2019] Generalized multiple importance sampling
+        https://arxiv.org/abs/1511.03095
+    """
+    # one-liner: mixture_u_n = -logsumexp(f_k - u_kn.T, b=N_k, axis=1)
```

Comment:
why not use one-liner?

Reply:
Unsure which style is more readable.

Included both extremes (one-liner, vs. expanding all steps) -- maybe something in between is preferable, which only expands a couple steps.

Reply:
makes sense! Thanks